src/uct/ib/rc/accel/rc_mlx5_common.c

```diff
@@ -107,35 +153,48 @@ unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface)
             srq->free_idx  = next_index;
         }
 
-        desc_map = ~seg->srq.ptr_mask & UCS_MASK(iface->tm.mp.num_strides);
-        ucs_for_each_bit(i, desc_map) {
-            UCT_TL_IFACE_GET_RX_DESC(&rc_iface->super.super, &rc_iface->rx.mp,
-                                     desc, goto out);
-
-            /* Set receive data segment pointer. Length is pre-initialized. */
-            hdr                = uct_ib_iface_recv_desc_hdr(&rc_iface->super, desc);
-            seg->srq.ptr_mask |= UCS_BIT(i);
-            seg->srq.desc      = desc; /* Optimization for non-MP case (1 stride) */
-            seg->dptr[i].lkey  = htonl(desc->lkey);
-            seg->dptr[i].addr  = htobe64((uintptr_t)hdr);
-            VALGRIND_MAKE_MEM_NOACCESS(hdr, rc_iface->super.config.seg_size);
+        if (uct_rc_mlx5_iface_srq_set_seg(iface, seg) != UCS_OK) {
+            break;
         }
 
         wqe_index = next_index;
     }
 
-out:
     count = wqe_index - srq->sw_pi;
-    ucs_assert(rc_iface->rx.srq.available >= count);
-
-    if (count > 0) {
-        srq->ready_idx              = wqe_index;
-        srq->sw_pi                  = wqe_index;
-        rc_iface->rx.srq.available -= count;
-        ucs_memory_cpu_store_fence();
-        *srq->db                    = htonl(srq->sw_pi);
-        ucs_assert(uct_ib_mlx5_srq_get_wqe(srq, srq->mask)->srq.next_wqe_index == 0);
+    uct_rc_mlx5_iface_update_srq_res(rc_iface, srq, wqe_index, count);
+    return count;
+}
+
+unsigned uct_rc_mlx5_iface_srq_post_recv_linked_list(
+        uct_rc_mlx5_iface_common_t *iface)
+{
+    uct_ib_mlx5_srq_t *srq   = &iface->rx.srq;
+    uct_rc_iface_t *rc_iface = &iface->super;
+    uct_ib_mlx5_srq_seg_t *seg = NULL;
+    uint16_t count = 0;
+    uint16_t wqe_index, next_index;
+
+    ucs_assert(rc_iface->rx.srq.available > 0);
+
+    wqe_index = srq->ready_idx;
+    seg = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);
+
+    for (;;) {
+        next_index = ntohs(seg->srq.next_wqe_index);
+        if (next_index == (srq->free_idx & srq->mask)) {
+            break;
+        }
+        seg = uct_ib_mlx5_srq_get_wqe(srq, next_index);
+
+        if (uct_rc_mlx5_iface_srq_set_seg(iface, seg) != UCS_OK) {
+            break;
+        }
+
+        wqe_index = next_index;
```

Comment:
move it outside the loop?

Reply:
unresolved

Reply:
unresolved: move it outside the loop?

Reply:
will yield value next to one we need

Reply:
is it possible to compute wqe_index outside the loop, from the value of next_index ?

Reply:
no. can try rewrite loop logic:
```
@@ -180,17 +180,17 @@ unsigned uct_rc_mlx5_iface_srq_post_recv_ll(uct_rc_mlx5_iface_common_t *iface)
     seg       = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);

     for (;;) {
-        next_index = ntohs(seg->srq.next_wqe_index);
-        if (next_index == (srq->free_idx & srq->mask)) {
+        if (ntohs(seg->srq.next_wqe_index) == (srq->free_idx & srq->mask)) {
             break;
         }
-        seg = uct_ib_mlx5_srq_get_wqe(srq, next_index);
+
+        wqe_index = ntohs(seg->srq.next_wqe_index);
+        seg       = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);

         if (uct_rc_mlx5_iface_srq_set_seg(iface, seg) != UCS_OK) {
             break;
         }

-        wqe_index = next_index;
         count++;
     }
```

Reply:
nvm